microbench/src/main/java/io/netty/microbench/concurrent/ScheduledFutureTaskBenchmark.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2017 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.microbench.concurrent;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.TimeUnit;
+
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.Level;
+import org.openjdk.jmh.annotations.Param;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.TearDown;
+import org.openjdk.jmh.infra.Blackhole;
+
+import io.netty.channel.DefaultEventLoop;
+import io.netty.channel.EventLoop;
+import io.netty.microbench.util.AbstractMicrobenchmark;
+import io.netty.util.concurrent.Future;
+import io.netty.util.concurrent.ScheduledFuture;
+
+@State(Scope.Benchmark)
+public class ScheduledFutureTaskBenchmark extends AbstractMicrobenchmark {
+
+    private static final EventLoop executor = new DefaultEventLoop();
+
+    @State(Scope.Thread)
+    public static class FuturesHolder {
+
+        @Param({"100", "1000", "10000", "100000"})
+        private int num;
+
+        private final List<ScheduledFuture<Void>> futures = new ArrayList<ScheduledFuture<Void>>();
+
+        @Setup(Level.Invocation)
+        public void reset() {
+            futures.clear();
+            for (int i = 1; i <= num; i++) {
+                futures.add(executor.schedule(new Callable<Void>() {
+                    @Override
+                    public Void call() throws Exception {
+                        return null;
+                    }
+                }, i, TimeUnit.HOURS));
+            }
+        }
+    }
+
+    @TearDown(Level.Trial)
+    public void stop() throws Exception {
+        executor.shutdownGracefully();
+    }
+
+    @Benchmark
+    public void cancelInOrder(Blackhole bh, final FuturesHolder futures) {
+        for (int i = 0; i < futures.num; i++) {
+            bh.consume(cancelFuture(futures.futures.get(i)));
+        }
+    }
+
+    @Benchmark
+    public void cancelInReverseOrder(Blackhole bh, final FuturesHolder futures) {
+        for (int i = futures.num - 1; i >= 0; i--) {
+            bh.consume(cancelFuture(futures.futures.get(i)));
```

Comment:
this benchmark has non trivial amount of overhead. suggestions:

- execute the test in the eventloop thread. executing a Runnable an blocking on the future is likely to have a high enough overhead to make the actual queue difference in the noise
- consider trying alternatives to Blackhole (e.g. return some value from the method that is the result of the computation).

Reply:
Can you give some pointers on running the test on the eventloop thread? Tried using `AbstractSharedExecutorMicrobenchmark` but it seemed to deadlock between the benchmark thread (JMH isn't asynchronous) and scheduler, which makes sense. Do you mean to use `AuxCounters`?

FWIW, even with the overhead, the benchmark results are pretty clear IMO.

Reply:
> Tried using AbstractSharedExecutorMicrobenchmark

The goal here is to supply JMH with the thread (an EventLoop) to run the benchmarks on. Its been a while since this was used, and I'm not surprised if there are issues here.

> Can you give some pointers on running the test on the eventloop thread?

The approach attempted by `AbstractSharedExecutorMicrobenchmark` would be ideal, but if there are issues with this approach then consider executing a runnable on the event loop which does all the work, and waiting for that to continue. This would reduce the context switching and Runnable allocation back/forth to just 1 per benchmark iteration instead of 1 for every schedule call.

> FWIW, even with the overhead, the benchmark results are pretty clear IMO.

I would expect the results to be more pronounced. We don't want to benchmark promise notification and context switching for each execution. Including "extras" in the benchmark makes the end results more volatile and more difficult to asses.

Reply:
Ah got it - thanks for the suggestion went with scheduling once per invocation.